File: frontend/types/analytics.ts
export type AnalyticsPreset =
  | "today"
  | "yesterday"
  | "last_7_days"
  | "last_30_days"
  | "this_month"
  | "last_month";

export type AnalyticsFormat = "csv" | "excel" | "pdf";

export type AnalyticsRange = {
  preset: AnalyticsPreset;
  start_date: string;
  end_date: string;
  timezone: string;
};

export type AnalyticsSeriesPoint = {
  label: string;
  value: string;
  count: number;
};

export type DashboardMetrics = {
  today_revenue: string;
  week_revenue: string;
  month_revenue: string;
  today_orders: number;
  week_orders: number;
  month_orders: number;
  average_invoice: string;
  total_products: number;
  low_stock_products: number;
  out_of_stock_products: number;
};

export type AnalyticsStatusBreakdown = {
  draft_orders: number;
  confirmed_orders: number;
  completed_orders: number;
  cancelled_orders: number;
  average_order_value: string;
  largest_order: string;
};

export type InventorySummary = {
  total_products: number;
  low_stock_products: number;
  out_of_stock_products: number;
  inventory_value: string;
};

export type TopProduct = {
  product_id: string | null;
  product_name: string;
  sku: string | null;
  unit: string | null;
  quantity_sold: string;
  revenue: string;
  order_count: number;
};

export type TopCategory = {
  category_name: string;
  quantity_sold: string;
  revenue: string;
  product_count: number;
};

export type TopCustomer = {
  customer_name: string;
  customer_phone: string | null;
  order_count: number;
  revenue: string;
  last_order_at: string;
};

export type RecentOrder = {
  id: string;
  invoice_number: string;
  customer_name: string | null;
  customer_phone: string | null;
  status: string;
  total: string;
  created_at: string;
};

export type DashboardResponse = {
  range: AnalyticsRange;
  metrics: DashboardMetrics;
  revenue_by_day: AnalyticsSeriesPoint[];
  orders_by_day: AnalyticsSeriesPoint[];
  top_products: TopProduct[];
  top_categories: TopCategory[];
  top_customers: TopCustomer[];
  recent_orders: RecentOrder[];
  inventory_summary: InventorySummary;
};

export type RevenueAnalyticsResponse = {
  range: AnalyticsRange;
  daily: AnalyticsSeriesPoint[];
  monthly: AnalyticsSeriesPoint[];
  metrics: DashboardMetrics;
};

export type OrdersAnalyticsResponse = {
  range: AnalyticsRange;
  daily: AnalyticsSeriesPoint[];
  monthly: AnalyticsSeriesPoint[];
  status_breakdown: AnalyticsStatusBreakdown;
  recent_orders: RecentOrder[];
};

export type ProductsAnalyticsResponse = {
  range: AnalyticsRange;
  top_products: TopProduct[];
  top_categories: TopCategory[];
  inventory_summary: InventorySummary;
};

export type CustomersAnalyticsResponse = {
  range: AnalyticsRange;
  top_customers: TopCustomer[];
  recent_orders: RecentOrder[];
};

File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


AnalyticsPreset = Literal["today", "yesterday", "last_7_days", "last_30_days", "this_month", "last_month"]
AnalyticsExportFormat = Literal["csv", "excel", "pdf"]
AnalyticsGranularity = Literal["day", "month"]


class AnalyticsRangeRead(BaseModel):
    preset: AnalyticsPreset
    start_date: date
    end_date: date
    timezone: str


class AnalyticsSeriesPointRead(BaseModel):
    label: str
    value: Decimal
    count: int = 0


class DashboardMetricsRead(BaseModel):
    today_revenue: Decimal
    week_revenue: Decimal
    month_revenue: Decimal
    today_orders: int
    week_orders: int
    month_orders: int
    average_invoice: Decimal
    total_products: int
    low_stock_products: int
    out_of_stock_products: int


class AnalyticsStatusBreakdownRead(BaseModel):
    new_orders: int
    confirmed_orders: int
    deleted_orders: int
    average_order_value: Decimal
    largest_order: Decimal


class InventorySummaryRead(BaseModel):
    total_products: int
    low_stock_products: int
    out_of_stock_products: int
    inventory_value: Decimal


class AnalyticsTopProductRead(BaseModel):
    product_id: UUID | None = None
    product_name: str
    sku: str | None = None
    unit: str | None = None
    quantity_sold: Decimal
    revenue: Decimal
    order_count: int


class AnalyticsTopCategoryRead(BaseModel):
    category_name: str
    quantity_sold: Decimal
    revenue: Decimal
    product_count: int


class AnalyticsTopCustomerRead(BaseModel):
    customer_name: str
    customer_phone: str | None = None
    order_count: int
    revenue: Decimal
    last_order_at: datetime


class AnalyticsRecentOrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    invoice_number: str
    customer_name: str | None = None
    customer_phone: str | None = None
    status: str
    total: Decimal
    created_at: datetime


class DashboardResponse(BaseModel):
    range: AnalyticsRangeRead
    metrics: DashboardMetricsRead
    revenue_by_day: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    orders_by_day: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    top_products: list[AnalyticsTopProductRead] = Field(default_factory=list)
    top_categories: list[AnalyticsTopCategoryRead] = Field(default_factory=list)
    top_customers: list[AnalyticsTopCustomerRead] = Field(default_factory=list)
    recent_orders: list[AnalyticsRecentOrderRead] = Field(default_factory=list)
    inventory_summary: InventorySummaryRead


class RevenueAnalyticsResponse(BaseModel):
    range: AnalyticsRangeRead
    daily: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    monthly: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    metrics: DashboardMetricsRead


class OrdersAnalyticsResponse(BaseModel):
    range: AnalyticsRangeRead
    daily: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    monthly: list[AnalyticsSeriesPointRead] = Field(default_factory=list)
    status_breakdown: AnalyticsStatusBreakdownRead
    recent_orders: list[AnalyticsRecentOrderRead] = Field(default_factory=list)


class ProductsAnalyticsResponse(BaseModel):
    range: AnalyticsRangeRead
    top_products: list[AnalyticsTopProductRead] = Field(default_factory=list)
    top_categories: list[AnalyticsTopCategoryRead] = Field(default_factory=list)
    inventory_summary: InventorySummaryRead


class CustomersAnalyticsResponse(BaseModel):
    range: AnalyticsRangeRead
    top_customers: list[AnalyticsTopCustomerRead] = Field(default_factory=list)
    recent_orders: list[AnalyticsRecentOrderRead] = Field(default_factory=list)


class AnalyticsExportRequest(BaseModel):
    format: AnalyticsExportFormat = "csv"
    preset: AnalyticsPreset = "last_30_days"
    start_date: date | None = None
    end_date: date | None = None
